isa/kernel.org/instruction-set.rst

```diff
@@ -143,57 +237,82 @@ order the operation convert from or to:
-  BPF_JA    0x00   PC += off                  BPF_JMP only
-  BPF_JEQ   0x10   PC += off if dst == src
-  BPF_JGT   0x20   PC += off if dst > src     unsigned
-  BPF_JGE   0x30   PC += off if dst >= src    unsigned
-  BPF_JSET  0x40   PC += off if dst & src
-  BPF_JNE   0x50   PC += off if dst != src
-  BPF_JSGT  0x60   PC += off if dst > src     signed
-  BPF_JSGE  0x70   PC += off if dst >= src    signed
-  BPF_CALL  0x80   function call
-  BPF_EXIT  0x90   function / program return  BPF_JMP only
-  BPF_JLT   0xa0   PC += off if dst < src     unsigned
-  BPF_JLE   0xb0   PC += off if dst <= src    unsigned
-  BPF_JSLT  0xc0   PC += off if dst < src     signed
-  BPF_JSLE  0xd0   PC += off if dst <= src    signed
-  ========  =====  =========================  ============
-
-The eBPF program needs to store the return value into register R0 before doing a
-BPF_EXIT.
 
+Support for ``BPF_JMP32`` is required in ISA version 3, and optional in earlier
+versions.
+
+The 4-bit 'code' field encodes the operation as below, where PC is the program counter:
+
+  ========  =====  ============================  =======  ============
+  code      value  description                   version  notes
+  ========  =====  ============================  =======  ============
+  BPF_JA    0x00   PC += offset                  1        BPF_JMP only
+  BPF_JEQ   0x10   PC += offset if dst == src    1
+  BPF_JGT   0x20   PC += offset if dst > src     1        unsigned
+  BPF_JGE   0x30   PC += offset if dst >= src    1        unsigned
+  BPF_JSET  0x40   PC += offset if dst & src     1
+  BPF_JNE   0x50   PC += offset if dst != src    1
+  BPF_JSGT  0x60   PC += offset if dst > src     1        signed
+  BPF_JSGE  0x70   PC += offset if dst >= src    1        signed
+  BPF_CALL  0x80   call function imm             1        see `Helper functions`_
+  BPF_EXIT  0x90   function / program return     1        BPF_JMP only
+  BPF_JLT   0xa0   PC += offset if dst < src     2        unsigned
+  BPF_JLE   0xb0   PC += offset if dst <= src    2        unsigned
+  BPF_JSLT  0xc0   PC += offset if dst < src     2        signed
+  BPF_JSLE  0xd0   PC += offset if dst <= src    2        signed
+  ========  =====  ============================  =======  ============
+
+where 'version' indicates the first ISA version in which the value was supported.
+
+The eBPF verifier is responsible for verifying that the
+eBPF program stores the return value into register R0 before doing a
+``BPF_EXIT``.
+
+Helper functions
+~~~~~~~~~~~~~~~~
+Helper functions are a concept whereby BPF programs can call into
+set of function calls exposed by the eBPF runtime.  Each helper
+function is identified by an integer used in a ``BPF_CALL`` instruction.
+The available helper functions may differ for each eBPF program type.
+
+Each helper function is implemented with a commonly shared function
+signature defined as:
+
+  uint64_t function(uint64_t r1, uint64_t r2, uint64_t r3, uint64_t r4, uint64_t r5)
```

Comment:
FYI, the above syntax comes from https://docs.cilium.io/en/stable/bpf/#helper-functions

I will see if I can add a sentence.  I think conceptually all five registers are passed and some arguments are ignored.   (The implementation details vary but I think the Isovalent doc mentioned isn't incorrect.)

Reply:
Yes I think it's technically correct when considering the low-level parts, I'm just thinking this might be confusing to readers who are only familiar with the existing BPF helpers

Reply:
Ok, added a paragraph after this line, please check.